### PR TITLE
Static deepcopy – phase 2: replace cloner calls by direct DeepCopy calls

### PR DESCRIPTION
**Pull request by sttts**
Based on https://github.com/kubernetes/kubernetes/pull/48544.

### CODE REVIEW

**Comment by sttts** [comment]
@mbohlool have you seen this before: the openapi json output is in different order?

```
I0725 21:35:28.674] @@ -1093,8 +1093,8 @@
I0725 21:35:28.674]       "x-kubernetes-action": "put",
I0725 21:35:28.675]       "x-kubernetes-group-version-kind": {
I0725 21:35:28.675]        "group": "",
I0725 21:35:28.675] -      "kind": "ConfigMap",
I0725 21:35:28.675] -      "version": "v1"
I0725 21:35:28.675] +      "version": "v1",
I0725 21:35:28.675] +      "kind": "ConfigMap"
I0725 21:35:28.675]       }
I0725 21:35:28.675]      },
```

**Comment by mbohlool** [comment]
@sttts Yes, that happened sometime. when you serialize/deserialize json objects. I didn't spend time on figuring out when/how exactly it happens.

File: federation/pkg/federation-controller/ingress/ingress_controller.go [hunk 1]
```diff
@@ -830,14 +804,7 @@ func (ic *IngressController) reconcileIngress(ingress types.NamespacedName) {
...
 				if !baseLBStatusExists && clusterLBStatusExists {
-					lbstatusObj, lbErr := api.Scheme.DeepCopy(&clusterIngress.Status.LoadBalancer)
-					lbstatus, ok := lbstatusObj.(*v1.LoadBalancerStatus)
-					if lbErr != nil || !ok {
-						glog.Errorf("Internal error: Failed to clone LoadBalancerStatus of %q in cluster %q while attempting to update master loadbalancer ingress status, will try again later. error: %v, Object to be cloned: %v", ingress, cluster.Name, lbErr, lbstatusObj)
-						ic.deliverIngress(ingress, ic.ingressReviewDelay, true)
-						return
-					}
-					baseIngress.Status.LoadBalancer = *lbstatus
+					clusterIngress.Status.LoadBalancer.DeepCopyInto(&baseIngress.Status.LoadBalancer)
```
Comment (deads2k) [inline federation/pkg/federation-controller/ingress/ingress_controller.go]: Why is this better than `baseIngress.Status.LoadBalancer = *clusterIngress.Status.LoadBalancer.DeepCopy()`?

File: federation/pkg/federation-controller/ingress/ingress_controller.go [hunk 2]
```diff
@@ -747,24 +736,9 @@ func (ic *IngressController) reconcileIngress(ingress types.NamespacedName) {
...
-		if !ok {
-			glog.Errorf("Internal error: Failed to cast to *metav1.ObjectMeta: %v", objMeta)
-		}
-		desiredIngress.ObjectMeta = *objMetaCopy
-		objSpecCopy, ok := objSpec.(*extensionsv1beta1.IngressSpec)
-		if !ok {
-			glog.Errorf("Internal error: Failed to cast to extensionsv1beta1.Ingressespec: %v", objSpec)
-		}
-		desiredIngress.Spec = *objSpecCopy
+		baseIngress.ObjectMeta.DeepCopyInto(&desiredIngress.ObjectMeta)
```
Comment (deads2k) [inline federation/pkg/federation-controller/ingress/ingress_controller.go]: similar question about just assigning directly.  Do we have things with handles to `ObjectMeta` at this stage we're worried about?

Comment (deads2k) [inline federation/pkg/federation-controller/ingress/ingress_controller.go]: Based on the old code, we don't.  I think the assignment is more clear to readers.  I'd pay for the allocation unless we find a problem.

File: federation/pkg/federation-controller/service/servicecontroller.go [hunk 3]
```diff
@@ -546,7 +537,7 @@ func getOperationsToPerformOnCluster(informer fedutil.FederatedInformer, cluster
 
 	desiredService := &v1.Service{
 		ObjectMeta: fedutil.DeepCopyRelevantObjectMeta(fedService.ObjectMeta),
-		Spec:       *(fedutil.DeepCopyApiTypeOrPanic(&fedService.Spec).(*v1.ServiceSpec)),
+		Spec:       *fedService.Spec.DeepCopy(),
```
Comment (deads2k) [inline federation/pkg/federation-controller/service/servicecontroller.go]: existing code is weird.  a copy of a copy?

File: pkg/kubelet/config/config.go [hunk 4]
```diff
@@ -503,11 +503,7 @@ func (s *podStorage) MergedState() interface{} {
 	pods := make([]*v1.Pod, 0)
 	for _, sourcePods := range s.pods {
 		for _, podRef := range sourcePods {
-			pod, err := api.Scheme.Copy(podRef)
-			if err != nil {
-				glog.Errorf("unable to copy pod: %v", err)
-			}
-			pods = append(pods, pod.(*v1.Pod))
+			pods = append(pods, podRef.DeepCopy())
```
Comment (deads2k) [inline pkg/kubelet/config/config.go]: this pull just keeps going.......

File: pkg/registry/rbac/reconciliation/rolebinding_interfaces.go [hunk 5]
```diff
@@ -26,6 +26,9 @@ import (
 	"k8s.io/kubernetes/pkg/client/clientset_generated/internalclientset/typed/rbac/internalversion"
 )
 
+// +k8s:deepcopy-gen=true
+// +k8s:deepcopy-gen:interfaces=k8s.io/kubernetes/pkg/registry/rbac/reconciliation.RoleBinding
+// +k8s:deepcopy-gen:nonpointer-interfaces=true
```
Comment (deads2k) [inline pkg/registry/rbac/reconciliation/rolebinding_interfaces.go]: interesting

**Comment by deads2k** [comment]
You know what.  I'll give you the nits.  Pass verify, squash in bazel and 

/lgtm

File: pkg/registry/rbac/reconciliation/rolebinding_interfaces.go (hunk 5 above)
Comment (sttts) [inline pkg/registry/rbac/reconciliation/rolebinding_interfaces.go]: Yes, we actually have an interface which does not hold a pointer.

File: federation/pkg/federation-controller/ingress/ingress_controller.go (hunk 1 above)
Comment (sttts) [inline federation/pkg/federation-controller/ingress/ingress_controller.go]: semantically the same, only saves a couple of allocations.

**Comment by sttts** [comment]
@deads2k addressed your nits.

@smarterclayton can you approve?

File: pkg/kubelet/kubelet_resources.go [hunk 6]
```diff
@@ -44,28 +43,14 @@ func (kl *Kubelet) defaultPodLimitsForDownwardAPI(pod *v1.Pod, container *v1.Con
...
 	if container != nil {
-		containerCopy, err := scheme.Scheme.DeepCopy(container)
-		if err != nil {
-			return nil, nil, fmt.Errorf("failed to perform a deep copy of container object: %v", err)
-		}
-		outputContainer, ok = containerCopy.(*v1.Container)
-		if !ok {
-			return nil, nil, fmt.Errorf("unexpected type returned from deep copy of container object")
-		}
+		outputContainer := container.DeepCopy()
```
Comment (sttts) [inline pkg/kubelet/kubelet_resources.go]: one hour of staring on logs and code later: the colon is not a good idea.

**Comment by sttts** [comment]
All single PRs for the several packages are merged. Next step: phase 3 https://github.com/kubernetes/kubernetes/pull/49575

**Pull request closed without merging**